zerver/tests/test_bots.py

```diff
@@ -414,6 +416,35 @@ def test_bot_add_subscription(self) -> None:
         # Test runner automatically redirects all sent email to a dummy 'outbox'.
         self.assert_length(mail.outbox, 0)
 
+    def test_bot_update_user_group(self) -> None:
+        self.login("hamlet")
+        user_group = UserGroup.objects.get(name="hamletcharacters")
+
+        # Create a bot.
+        self.assert_num_bots_equal(0)
+        bot = self.create_bot()
+        self.assert_num_bots_equal(1)
+
+        params = {"add": orjson.dumps([bot["user_id"]]).decode()}
+        initial_last_message = self.get_last_message()
+        result = self.client_post(f"/json/user_groups/{user_group.id}/members", info=params)
+        self.assert_json_success(result)
+
+        self.assertEqual(UserGroupMembership.objects.filter(user_group=user_group).count(), 3)
+
+        # No notification message is sent for adding a bot
+        self.assertEqual(self.get_last_message(), initial_last_message)
+
+        params = {"delete": orjson.dumps([bot["user_id"]]).decode()}
+        initial_last_message = self.get_last_message()
+        result = self.client_post(f"/json/user_groups/{user_group.id}/members", info=params)
+        self.assert_json_success(result)
+
+        self.assertEqual(UserGroupMembership.objects.filter(user_group=user_group).count(), 2)
+
+        # No notification message is sent for removing a bot
+        self.assertEqual(self.get_last_message(), initial_last_message)
```

Comment:
I find this sort of test tends to not be very effective, because it is separated from the non-bot case. 

I think we should delete this test, and just extend `test_update_members_of_user_group` to add/remove some bot users in the same API request as non-bot users, so that we're doing the assertions in a code path that also has notifications being sent to other users. (I think the same idea will work for deactivated users)

See https://zulip.readthedocs.io/en/latest/testing/philosophy.html#share-test-setup-code for some more discussion (the rest of that page is probably good reading too!).